File: tools/tapscore-watch/WatchApp/MatchModel.swift
import Foundation
import Observation
import TapScoreEngine

/// Owns the live match, drives the shared engine, plays haptics, and remembers the last format.
/// Scoring rules are NOT here — they live in TapScoreEngine (single source of truth).
/// Uses the @Observable macro (watchOS 10+) — views observe reads automatically, no Combine.
@Observable
final class MatchModel {
    private(set) var match: MatchState
    var active: Bool = false                      // false = show Start screen; true = scoring

    @ObservationIgnored private var history: [MatchState] = []   // undo stack (snapshots); not observed
    private static let settingsKey = "tapscore.lastSettings"

    init() {
        match = MatchState(settings: MatchModel.loadSettings())
    }

    // MARK: derived (for views)

    var pointLabels: [String] { ScoringEngine.pointDisplay(match) }
    var canUndo: Bool { !history.isEmpty }

    /// Compact set line, e.g. "SET 1-0" (racket) or "GAME 1-0" (target sports).
    var setsLine: String {
        let noun = ScoringEngine.isTargetSport(match.settings.sport) ? "GAME" : "SET"
        return "\(noun) \(match.sets[0])-\(match.sets[1])"
    }

    // MARK: actions

    func score(_ side: Int) {
        guard !match.over else { return }
        history.append(match)
        let before = match
        ScoringEngine.scorePoint(&match, side)
        if match.over {
            Haptics.matchWon()
        } else if setsTotal(match) > setsTotal(before) {
            Haptics.setWon()
        } else if gamesTotal(match) > gamesTotal(before) {
            Haptics.game()
        } else {
            Haptics.point()
        }
    }

    func undo() {
        guard let prev = history.popLast() else { return }
        match = prev
        Haptics.undo()
    }

    func startMatch(_ settings: Settings) {
        MatchModel.saveSettings(settings)
        match = MatchState(settings: settings)
        history = []
        active = true
    }

    /// Leave the current match and return to Start (no result kept).
    func endMatch() {
        active = false
    }

    /// Start a fresh match with the same settings (used from the winner screen).
    func rematch() { startMatch(match.settings) }

    // MARK: helpers

    private func setsTotal(_ m: MatchState) -> Int { m.sets[0] + m.sets[1] }
    private func gamesTotal(_ m: MatchState) -> Int { m.games[0] + m.games[1] }

    // MARK: persistence (remember the last-used format for instant standalone start)

    static func loadSettings() -> Settings {
        if let data = UserDefaults.standard.data(forKey: settingsKey),
           let s = try? JSONDecoder().decode(Settings.self, from: data) {
            return s
        }
        return Settings()   // tennis, best of 3
    }

    static func saveSettings(_ s: Settings) {
        if let data = try? JSONEncoder().encode(s) {
            UserDefaults.standard.set(data, forKey: settingsKey)
        }
    }
}

File: tools/tapscore-wear/WearApp/MatchModel.kt
package com.ivogomes.tapscore.wear

import android.content.Context
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.setValue
import com.ivogomes.tapscore.engine.MatchState
import com.ivogomes.tapscore.engine.ScoringEngine
import com.ivogomes.tapscore.engine.Settings

/**
 * Owns the live match, drives the shared engine, plays haptics, and remembers the last format.
 * Scoring rules are NOT here — they live in ScoringEngine (single source of truth).
 *
 * Holds Compose state directly (`mutableStateOf`), so views recompose when `match`/`active` change.
 * Each score/undo swaps in a fresh MatchState instance so recomposition triggers reliably.
 */
class MatchModel(context: Context) {
    private val prefs = context.applicationContext.getSharedPreferences("tapscore", Context.MODE_PRIVATE)
    private val haptics = Haptics(context)

    var match by mutableStateOf(MatchState(loadSettings()))
        private set
    var active by mutableStateOf(false)          // false = Start screen; true = scoring
        private set

    private val history = ArrayDeque<MatchState>()   // undo stack (snapshots)

    val pointLabels: List<String> get() = ScoringEngine.pointDisplay(match)
    val canUndo: Boolean get() = history.isNotEmpty()

    /** Full scoreline for the center pill (completed sets + current games), like the phone board. */
    val scorePill: String
        get() {
            val m = match
            val parts = m.completedSets.map { "${it[0]}-${it[1]}" }.toMutableList()
            if (!ScoringEngine.isTargetSport(m.settings.sport)) {
                parts.add("${m.games[0]}-${m.games[1]}")   // current set's games
            }
            if (parts.isEmpty()) parts.add("0-0")
            return parts.joinToString("  ")
        }

    // MARK: actions

    fun score(side: Int) {
        if (match.over) return
        history.addLast(match)                   // snapshot current (never mutated again)
        val before = match
        val next = match.deepCopy()
        ScoringEngine.scorePoint(next, side)
        when {
            next.over -> haptics.matchWon()
            setsTotal(next) > setsTotal(before) -> haptics.setWon()
            gamesTotal(next) > gamesTotal(before) -> haptics.game()
            else -> haptics.point()
        }
        match = next
    }

    fun undo() {
        val prev = history.removeLastOrNull() ?: return
        match = prev
        haptics.undo()
    }

    fun startMatch(settings: Settings) {
        saveSettings(settings)
        match = MatchState(settings)
        history.clear()
        active = true
    }

    /** Leave the current match, back to Start (no result kept). */
    fun endMatch() { active = false }

    /** Fresh match with the same settings (from the winner screen). */
    fun rematch() = startMatch(match.settings)

    private fun setsTotal(m: MatchState) = m.sets[0] + m.sets[1]
    private fun gamesTotal(m: MatchState) = m.games[0] + m.games[1]

    // MARK: persistence (remember last-used format for instant standalone start)

    fun loadSettings(): Settings {
        val sport = prefs.getString("sport", "tennis") ?: "tennis"
        val bestOf = prefs.getInt("bestOf", 3)
        val s = Settings()
        s.sport = sport
        s.bestOf = bestOf
        if (ScoringEngine.isTargetSport(sport)) s.pointsTarget = 11
        return s
    }

    private fun saveSettings(s: Settings) {
        prefs.edit()
            .putString("sport", s.sport)
            .putInt("bestOf", s.bestOf ?: 3)
            .apply()
    }
}
